src/monitor/pgautofailover--1.5--1.6.sql

```diff
@@ -405,6 +454,91 @@ CREATE TRIGGER disable_secondary_check
 	EXECUTE PROCEDURE pgautofailover.update_secondary_check();
 
 
+CREATE FUNCTION pgautofailover.last_events
+ (
+  count int default 10
+ )
+RETURNS SETOF pgautofailover.event LANGUAGE SQL STRICT
+AS $$
+with last_events as
+(
+  select eventid, eventtime, formationid,
+         nodeid, groupid, nodename, nodehost, nodeport,
+         reportedstate, goalstate,
+         reportedrepstate, reportedtli, reportedlsn,
+         candidatepriority, replicationquorum, description
+    from pgautofailover.event
+order by eventid desc
+   limit count
+)
+select * from last_events order by eventtime, eventid;
+$$;
+
+comment on function pgautofailover.last_events(int)
+        is 'retrieve last COUNT events';
+
+grant execute on function pgautofailover.last_events(int)
+   to autoctl_node;
+
+CREATE FUNCTION pgautofailover.last_events
+ (
+  formation_id text default 'default',
+  count        int  default 10
+ )
+RETURNS SETOF pgautofailover.event LANGUAGE SQL STRICT
+AS $$
+with last_events as
+(
+    select eventid, eventtime, formationid,
+           nodeid, groupid, nodename, nodehost, nodeport,
+           reportedstate, goalstate,
+           reportedrepstate, reportedtli, reportedlsn,
+           candidatepriority, replicationquorum, description
+      from pgautofailover.event
+     where formationid = formation_id
+  order by eventid desc
+     limit count
+)
+select * from last_events order by eventtime, eventid;
+$$;
+
+comment on function pgautofailover.last_events(text,int)
+        is 'retrieve last COUNT events for given formation';
+
+grant execute on function pgautofailover.last_events(text,int)
+   to autoctl_node;
+
+CREATE FUNCTION pgautofailover.last_events
+ (
+  formation_id text,
+  group_id     int,
+  count        int default 10
+ )
+RETURNS SETOF pgautofailover.event LANGUAGE SQL STRICT
+AS $$
+with last_events as
+(
+    select eventid, eventtime, formationid,
+           nodeid, groupid, nodename, nodehost, nodeport,
+           reportedstate, goalstate,
+           reportedrepstate, reportedtli, reportedlsn,
+           candidatepriority, replicationquorum, description
+      from pgautofailover.event
+     where formationid = formation_id
+       and groupid = group_id
+  order by eventid desc
+     limit count
+)
+select * from last_events order by eventtime, eventid;
+$$;
+
+comment on function pgautofailover.last_events(text,int,int)
+        is 'retrieve last COUNT events for given formation and group';
+
+grant execute on function pgautofailover.last_events(text,int,int)
+   to autoctl_node;
```

Comment:
All these grants are actualy not in `pgautofailover.sql`, I guess it doesn't need access to these functions.

Reply:
Other grants on functions are edited the same (only in the upgrade script), and I came to the theory that the upgrade script is run by the superuser and default grants would not be adequate?

Reply:
Actually most functions have the grant execute lines, no reason to bypass this one. Added now.